perf: synchronous updates

wSedlacek opened the pull request:
This refactors `useTextWidth` to make use of `useMemo` so that updates are synchronous.
Additionally this moves `getTextWidth` and `getContext` to the module scope so that they do not need to be redeclared on every effect call.

fixes: #1

Commit:
perf: synchronous updates

This refactors `useTextWidth` to make use of `useMemo` so that updates are synchronous.
Additionally this moves `getTextWidth` and `getContext` to the module scope so that they do not need to be redeclared on every effect call.

erhangundogan reviewed (commented):
@wSedlacek thanks a lot for the changes. As you mentioned there is no need to use state. Hook can return width directly. 

Just one thing, there are a couple of lint errors in your code. Could you please run linter to see errors and fix them OR you can fix them automatically with command below?
```
yarn lint --fix
```

Commit:
refactor: strict definitions

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

wSedlacek commented:
I noticed we had a few linting warnings as well.
I went ahead and fixed those as well as enabling strict mode in TypeScript to further protect against null reference errors.

Update: I am going to do a few other fixes to improve the interface.

Commit:
feat!: improve options interface

BREAKING CHANGE: This requires `text` OR `ref` to be defined both via the TypeScript interface and at runtime.

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

wSedlacek commented:
I'll leave it at this for you to decide if you want to do anything else with this.

Some notes,
~~I noticed we don't have any `ref` test.~~
~~I noticed we don't have any test for assuring updates are applied.~~
There are some breaking changes,
- with the interface requiring either `ref` or `text` now, an error will be thrown if both are undefined (previously undefined was returned which was not type safe)
- when `ref.current` is `null` `NaN` will be returned (rather than undefined previously which was not type safe)

Commit:
fix: restore memorization deps

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

Commit:
style: fix linting errors

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

Commit:
test: element ref object test

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

wSedlacek commented:
🎉 100% test coverage

Commit:
test: hook updates

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

erhangundogan commented on src/useTextWidth.tsx:
```diff
@@ -1,44 +1,57 @@
-  }, [font, ref, text]);
 
-  return width;
+    throw new TypeError('useTextWidth - Either `ref` OR `text` must be defined');
```
I think this might cause unwanted error messages in the console when the text/ref assigned but the actual value is still undefined or null. This is a common case during the application init phase. Also people should be aware of the hook argument required here. Throwing exception won't help for good IMO.

wSedlacek replied:
Ref's should still have a `.current` even if they haven't been assigned, so they wouldn't encounter this unless there is some funky manipulation going on.

As far as Text, in strict mode projects they would be warned by the interface that `undefined` is not a valid value for `text`
That said we could loosen up that and return NaN like I have setup for when `ref.current` is null.

erhangundogan reviewed (commented):
@wSedlacek thanks a lot for the improvements, typedefs and code coverage. It looks much better now :) 

I just added one comment for throwing exception. Apart from that rest is good.

Commit:
refactor: fix strict mode example

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

Commit:
feat: support undefined text

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

erhangundogan commented:
@wSedlacek I think we can bump minor version in package.json wdyt? You can change it to `1.1.0` and then I can create tag after merge. I was doing this semantic versioning manually.

Commit:
chore: version bump 1.1

Signed-off-by: wSedlacek <wsedlacekc@gmail.com>

wSedlacek commented:
When can I expect this to be published to NPM?

erhangundogan commented:
I have just published. Sorry for the delay.
